flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/RestartBackoffTimeStrategyFactoryLoader.java

```diff
@@ -0,0 +1,393 @@
+				jobRestartStrategyConfiguration,
+				clusterConfiguration,
+				isCheckpointingEnabled);
+
+			// create new restart strategy from the adapted config
+			return createRestartStrategyFactoryInternal(adaptedConfiguration);
+		}
+	}
+
+	/**
+	 * Decides the {@link RestartBackoffTimeStrategy} to use and its params based on legacy configs,
+	 * and records its class name and the params into a adapted configuration.
+	 *
+	 * <p>The decision making is as follows:
+	 * <ol>
+	 *     <li>Use strategy of {@link RestartStrategies.RestartStrategyConfiguration}, unless it's a
+	 * {@link RestartStrategies.FallbackRestartStrategyConfiguration} or not valid.</li>
+	 *     <li>If strategy is not decided, use legacy strategy specified in cluster(server-side) config,
+	 * unless it is not set or not valid</li>
+	 *     <li>If strategy is not decided, use {@link FixedDelayRestartStrategy} if checkpointing is
+	 * enabled. Otherwise use {@link NoRestartStrategy}</li>
+	 * </ol>
+	 *
+	 * @param jobRestartStrategyConfiguration restart configuration given within the job graph
+	 * @param clusterConfiguration cluster(server-side) configuration, usually represented as jobmanager config
+	 * @param isCheckpointingEnabled if checkpointing was enabled for the job
+	 * @return adapted configuration
+	 */
+	private static Configuration getAdaptedConfiguration(
+			final RestartStrategies.RestartStrategyConfiguration jobRestartStrategyConfiguration,
+			final Configuration clusterConfiguration,
+			final boolean isCheckpointingEnabled) throws Exception {
+
+		// clone a new config to not modify existing one
+		final Configuration adaptedConfiguration = new Configuration();
+
+		// try determining restart strategy based on job restart strategy config first
+		if (!tryAdaptingJobRestartStrategyConfiguration(jobRestartStrategyConfiguration, adaptedConfiguration)) {
+
+			// if job restart strategy config does not help
+			// try determining restart strategy based on cluster config as fallback
+			if (!tryAdaptingClusterRestartStrategyConfiguration(clusterConfiguration, adaptedConfiguration)) {
+
+				// if the restart strategy is not decided yet
+				// determine the strategy based on whether checkpointing is enabled
+				setDefaultRestartStrategyToConfiguration(isCheckpointingEnabled, adaptedConfiguration);
+			}
+		}
+
+		return adaptedConfiguration;
+	}
+
+	private static RestartBackoffTimeStrategy.Factory createRestartStrategyFactoryInternal(
+			final Configuration configuration) {
+
+		final String restartStrategyClassName = configuration.getString(
+			RestartBackoffTimeStrategyOptions.RESTART_BACKOFF_TIME_STRATEGY_CLASS_NAME);
+
+		if (restartStrategyClassName == null) {
+			LOG.info("No restart strategy is configured. Using no restart strategy.");
+			return new NoRestartBackoffTimeStrategy.NoRestartBackoffTimeStrategyFactory();
+		}
+
+		try {
+			final Class<?> clazz = Class.forName(restartStrategyClassName);
```

Comment:
If we do not offer users to provide their own strategy (see one of my other comments), we can rethink whether it is _"correct"_ to always instantiate the strategies by reflection.

Reply:
Yes. I will remove the customization logic from this PR. 
Will add it in another PR if the discussion result is that customizable restart strategies are still required.

Reply:
As long as there are unit tests, it is probably reasonably safe to keep instantiating by reflection.

Reply:
Actually I changed my mind on this matter. If we remove support for custom restart strategies/back-off time strategies, there is no apparent reason to instantiate the strategy reflectively. I would propose to do the simplest thing possible to avoid over-engineering the code. Also the fact that we require reflective invocation of the `createFactory` method means to me that something is wrong. Should we need custom strategies, we can consider one of the following alternatives for loading them:

- http://mail-archives.apache.org/mod_mbox/flink-dev/201904.mbox/%3C02763EDD-4806-4649-91ED-D7E8205C5D64@ververica.com%3E
- https://docs.oracle.com/javase/9/docs/api/java/util/ServiceLoader.html